pkg/virt-controller/watch/vm.go

```diff
@@ -2025,6 +1921,130 @@ func (c *VMController) processFailureCondition(vm *virtv1.VirtualMachine, vmi *v
+		stateChange := vm.Status.StateChangeRequests[0]
+		switch stateChange.Action {
+		case virtv1.StopRequest:
+			if vmi == nil {
+				// because either the VM or VMI informers can trigger processing here
+				// double check the state of the cluster before taking action
+				_, err := c.clientset.VirtualMachineInstance(vm.ObjectMeta.Namespace).Get(vm.GetName(), &v1.GetOptions{})
+				if err != nil && errors.IsNotFound(err) {
+					// If there's no VMI, then the VMI was stopped, and the stopRequest can be cleared
+					log.Log.Object(vm).V(4).Infof("No VMI. Clearing stop request")
+					clearChangeRequest = true
+				}
+			} else {
+				if stateChange.UID == nil {
+					// It never makes sense to have a request to stop a VMI that doesn't
+					// have a UUID associated with it. This shouldn't be possible -- but if
+					// it occurs, clear the stopRequest because it can't be acted upon
+					log.Log.Object(vm).Errorf("Stop Request has no UID.")
+					clearChangeRequest = true
+				} else if *stateChange.UID != vmi.UID {
+					// If there is a VMI, but the UID doesn't match, then it
+					// must have been previously stopped, so the stopRequest can be cleared
+					log.Log.Object(vm).V(4).Infof("VMI's UID doesn't match. clearing stop request")
+					clearChangeRequest = true
+				}
+			}
+		case virtv1.StartRequest:
+			// If the current VMI is running, then it has been started.
+			if vmi != nil {
+				log.Log.Object(vm).V(4).Infof("VMI exists. clearing start request")
+				clearChangeRequest = true
+			}
+		}
+	}
+
+	return clearChangeRequest
+}
+
+func (c *VMController) trimDoneVolumeRequests(vm *virtv1.VirtualMachine) {
+	if len(vm.Status.VolumeRequests) > 0 {
+		volumeMap := make(map[string]virtv1.Volume)
+		diskMap := make(map[string]virtv1.Disk)
+
+		for _, volume := range vm.Spec.Template.Spec.Volumes {
+			volumeMap[volume.Name] = volume
+		}
+		for _, disk := range vm.Spec.Template.Spec.Domain.Devices.Disks {
+			diskMap[disk.Name] = disk
+		}
+
+		tmpVolRequests := vm.Status.VolumeRequests[:0]
+		for _, request := range vm.Status.VolumeRequests {
+
+			var added bool
+			var volName string
+
+			removeRequest := false
+
+			if request.AddVolumeOptions != nil {
+				volName = request.AddVolumeOptions.Name
+				added = true
+			} else if request.RemoveVolumeOptions != nil {
+				volName = request.RemoveVolumeOptions.Name
+				added = false
+			}
+
+			_, volExists := volumeMap[volName]
+			_, diskExists := diskMap[volName]
+
+			if added && volExists && diskExists {
+				removeRequest = true
+			} else if !added && !volExists && !diskExists {
+				removeRequest = true
+			}
+
+			if !removeRequest {
+				tmpVolRequests = append(tmpVolRequests, request)
+			}
+		}
+		vm.Status.VolumeRequests = tmpVolRequests
+	}
+}
+
+func (c *VMController) handleScalingUpOrDown(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance, key string, dataVolumes []*cdiv1.DataVolume) (syncError, error) {
```

Comment:
I'm not sure `handleScalingUpOrDown` is the best name for this function...

What this function is doing is synchronizing the ChangeRequest and indirectly it can scale up or down, starting or stopping a VMI.

Also, note that when we introduce tracing in this code, we call it `sync`, to measure the operating time of `sync`.

So I suggest calling this function like `syncChangeRequest()`